### ISSUE TITLE
exception in train.py 

### ISSUE THREAD

**Issue by deyituo** [zhaojw1998/Beat-Transformer#2]
It seems like that sometimes the train.py will throw exception when calculating loss? 

Loss.cu:102: operator(): block: [0,0,0], thread: [6,0,0] Assertion input_val >= zero && input_val <= one failed.

**Comment by zhaojw1998** [comment]
I came across this error too. It seems to be some CUDA kernel problem and it happens randomly, and it may not occur as well. I did not make it to debug though. I simply ran train.py several times in parallel and there can hopefully be a successful one without error.

**Comment by deyituo** [comment]
I go through the code and find that 2 reasons could cause this exception: 

1. the label of gt/tempo_gt include -1 for masking, fixed version:

```
            valid_gt = gt.clone()
            valid_gt[gt == -1] = 0
            loss = loss_func(pred, valid_gt)
            weight = (1 - torch.as_tensor(gt == -1, dtype=torch.int32)).to(device)
            loss = (weight * loss).mean(dim=(0, 1)).sum()

            valid_tempo_gt = tempo_gt.clone()
            valid_tempo_gt[tempo_gt == -1] = 0
            loss_t = loss_tempo(torch.softmax(tempo, dim=-1), valid_tempo_gt)
            weight = (1 - torch.as_tensor(tempo_gt == -1, dtype=torch.int32)).to(device)
            loss_t = (weight * loss_t).mean()
```


2. return value and try/catch should be added to these two functions, fixed version:

```
def infer_beat_with_DBN(beat_pred, beat_gt, dbn_model, fps):
    #beat_pred: (B, L), estimation result
    weight = (1 - torch.as_tensor(beat_gt == -1, dtype=torch.int32))
    beat_pred = (torch.sigmoid(beat_pred) * weight).detach().cpu().numpy()
    #beat_pred = (beat_pred / fps)
    beat_gt = torch.as_tensor((beat_gt - 0.5) > 0, dtype=torch.int32).detach().cpu().numpy()
    batch_score = []
    for idx in range(beat_pred.shape[0]):
        #if (beat_gt[idx] == 0).all():
        #    continue
        if np.sum(beat_gt[idx]) < 2:
            continue
        try:
            beat_pred_batch = dbn_model(beat_pred[idx])
        except:
            return {"fmeasure": 0, "cmlt": 0, "amlt": 0}
        beat_gt_batch = np.nonzero(beat_gt[idx])[0] / fps
        score = madmom.evaluation.beats.BeatEvaluation(beat_pred_batch, beat_gt_batch)
        batch_score.append(score)
    batch_score = madmom.evaluation.beats.BeatMeanEvaluation(batch_score)
    return {"fmeasure": batch_score.fmeasure if not np.isnan(batch_score.fmeasure) else 0, \
            #"cemgil": batch_score.cemgil, \
            #"cmlc": batch_score.cmlc, \
            "cmlt": batch_score.cmlt if not np.isnan(batch_score.cmlt) else 0, \
            #"amlc": batch_score.amlc, \
            "amlt": batch_score.amlt if not np.isnan(batch_score.amlt) else 0}

def infer_downbeat_with_DBN(beat_pred, downbeat_pred, downbeat_gt, dbn_model, fps):
    #beat_pred: (B, L), estimation result
    beat_pred = torch.sigmoid(beat_pred).detach().cpu()
    downbeat_pred = torch.sigmoid(downbeat_pred).detach().cpu()
    combined_act = torch.cat((torch.maximum(beat_pred - downbeat_pred, torch.zeros(beat_pred.shape)).unsqueeze(-1), downbeat_pred.unsqueeze(-1)), dim=-1)
    #beat_pred = (beat_pred / fps)
    weight = (1 - torch.as_tensor(downbeat_gt == -1, dtype=torch.int32)).unsqueeze(-1).detach().cpu()
    combined_act = (combined_act * weight).numpy()

    beat_gt = torch.as_tensor((downbeat_gt - 0.5) > 0, dtype=torch.int32).detach().cpu().numpy()
    batch_score = []
    for idx in range(beat_pred.shape[0]):
        #if (beat_gt[idx] == 0).all():
        #    continue
        if np.sum(beat_gt[idx]) < 2:
            continue
        try:
            beat_pred_batch = dbn_model(combined_act[idx])
            beat_pred_batch = beat_pred_batch[beat_pred_batch[:, 1]==1][:, 0]
        except:
            return {"fmeasure": 0, "cmlt": 0, "amlt": 0}
        beat_gt_batch = np.nonzero(beat_gt[idx])[0] / fps
        score = madmom.evaluation.beats.BeatEvaluation(beat_pred_batch, beat_gt_batch)
        batch_score.append(score)
    batch_score = madmom.evaluation.beats.BeatMeanEvaluation(batch_score)
    return {"fmeasure": batch_score.fmeasure if not np.isnan(batch_score.fmeasure) else 0, \
            #"cemgil": batch_score.cemgil, \
            #"cmlc": batch_score.cmlc, \
            "cmlt": batch_score.cmlt if not np.isnan(batch_score.cmlt) else 0, \
            #"amlc": batch_score.amlc, \
            "amlt": batch_score.amlt if not np.isnan(batch_score.amlt) else 0}
```

**Comment by zhaojw1998** [comment]
Thank you very much! It helps a lot!
